File: infinity-server/src/main/java/io/github/derechtepilz/infinity/structure/StructureLoader.java
package io.github.derechtepilz.infinity.structure;

import com.google.gson.JsonArray;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import io.github.derechtepilz.infinity.Infinity;
import org.bukkit.Bukkit;
import org.bukkit.Material;
import org.bukkit.NamespacedKey;
import org.bukkit.World;
import org.bukkit.block.Block;
import org.bukkit.block.BlockFace;
import org.bukkit.block.BlockState;
import org.bukkit.block.data.Bisected;
import org.bukkit.block.data.BlockData;
import org.bukkit.block.data.Directional;
import org.bukkit.block.data.type.Leaves;
import org.bukkit.block.data.type.Slab;
import org.bukkit.block.data.type.Stairs;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;

public class StructureLoader {

	private final String world;
	private final InputStream structureFile;

	public StructureLoader(String world, InputStream structureFile) {
		this.world = world;
		this.structureFile = structureFile;

		generateStructure();
	}

	private void generateStructure() {
		try {
			World world = Bukkit.getWorld(new NamespacedKey(Infinity.NAME, this.world));
			BufferedReader structureReader = new BufferedReader(new InputStreamReader(structureFile));
			StringBuilder builder = new StringBuilder();
			String line;
			while ((line = structureReader.readLine()) != null) {
				builder.append(line);
			}
			JsonArray jsonArray = JsonParser.parseString(builder.toString()).getAsJsonArray();
			for (int i = 0; i < jsonArray.size(); i++) {
				JsonObject blockInformationJson = jsonArray.get(i).getAsJsonObject();
				int locX = blockInformationJson.get("locX").getAsInt();
				int locY = blockInformationJson.get("locY").getAsInt();
				int locZ = blockInformationJson.get("locZ").getAsInt();
				Material material = Material.matchMaterial(blockInformationJson.get("materialType").getAsString());
				world.setType(locX, locY, locZ, material);
				// Change BlockData
				Block placedBlock = world.getBlockAt(locX, locY, locZ);
				BlockState placedBlockState = placedBlock.getState();
				BlockData placedBlockData = placedBlockState.getBlockData();

				if (blockInformationJson.has("direction")) {
					Directional directional = (Directional) placedBlockData;
					directional.setFacing(BlockFace.valueOf(blockInformationJson.get("direction").getAsString()));
					placedBlockData = directional;
				}

				if (blockInformationJson.has("shapeType")) {
					Stairs shaped = (Stairs) placedBlockData;
					shaped.setShape(Stairs.Shape.valueOf(blockInformationJson.get("shapeType").getAsString()));
					placedBlockData = shaped;
				}

				if (blockInformationJson.has("bisectedHalf")) {
					Bisected bisected = (Bisected) placedBlockData;
					bisected.setHalf(Bisected.Half.valueOf(blockInformationJson.get("bisectedHalf").getAsString()));
					placedBlockData = bisected;
				}

				if (blockInformationJson.has("slabHalf")) {
					Slab slab = (Slab) placedBlockData;
					slab.setType(Slab.Type.valueOf(blockInformationJson.get("slabHalf").getAsString()));
					placedBlockData = slab;
				}

				if (blockInformationJson.has("persistent")) {
					Leaves leaves = (Leaves) placedBlockData;
					leaves.setPersistent(blockInformationJson.get("persistent").getAsBoolean());
					placedBlockData = leaves;
				}

				placedBlockState.setBlockData(placedBlockData);
				placedBlockState.update();
			}
		} catch (IOException e) {
			Infinity.getInstance().getLogger().severe("There has been a problem while reading resource files. Generated Infinity worlds may be empty. Please report this, alongside any helpful information like a full server log, etc.");
		}
	}

}

File: infinity-server/src/main/kotlin/io/github/derechtepilz/infinity/structure/StructureLoader.kt
/*
 *  Infinity - a Minecraft story-game for Paper servers
 *  Copyright (C) 2023  DerEchtePilz
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package io.github.derechtepilz.infinity.structure

import com.google.gson.JsonParser
import io.github.derechtepilz.infinity.Infinity0
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.NamespacedKey
import org.bukkit.block.BlockFace
import org.bukkit.block.data.Bisected
import org.bukkit.block.data.Bisected.Half
import org.bukkit.block.data.Directional
import org.bukkit.block.data.type.Leaves
import org.bukkit.block.data.type.Slab
import org.bukkit.block.data.type.Slab.Type
import org.bukkit.block.data.type.Stairs
import org.bukkit.block.data.type.Stairs.Shape
import java.io.BufferedReader
import java.io.InputStream
import java.io.InputStreamReader

class StructureLoader(private val world: String, private val structureFile: InputStream) {

	init {
		generateStructure()
	}

	private fun generateStructure() {
		val world = Bukkit.getWorld(NamespacedKey(Infinity0.NAME, world))!!
		val structureReader = BufferedReader(InputStreamReader(structureFile))
		val builder = java.lang.StringBuilder()
		var line: String?
		while (structureReader.readLine().also { line = it } != null) {
			builder.append(line)
		}
		val jsonArray = JsonParser.parseString(builder.toString()).asJsonArray
		for (i in 0 until jsonArray.size()) {
			val blockInformationJson = jsonArray[i].asJsonObject
			val locX = blockInformationJson.get("locX").asInt
			val locY = blockInformationJson.get("locY").asInt
			val locZ = blockInformationJson.get("locZ").asInt
			val material = Material.matchMaterial(blockInformationJson.get("materialType").asString)!!
			world.setType(locX, locY, locZ, material)
			// Change BlockData
			val placedBlock = world.getBlockAt(locX, locY, locZ)
			val placedBlockState = placedBlock.state
			var placedBlockData = placedBlockState.blockData

			if (blockInformationJson.has("direction")) {
				val directional: Directional = placedBlockData as Directional
				directional.facing = BlockFace.valueOf(blockInformationJson.get("direction").asString)
				placedBlockData = directional
			}

			if (blockInformationJson.has("shapeType")) {
				val shaped: Stairs = placedBlockData as Stairs
				shaped.shape = Shape.valueOf(blockInformationJson.get("shapeType").asString)
				placedBlockData = shaped
			}

			if (blockInformationJson.has("bisectedHalf")) {
				val bisected: Bisected = placedBlockData as Bisected
				bisected.half = Half.valueOf(blockInformationJson.get("bisectedHalf").asString)
				placedBlockData = bisected
			}

			if (blockInformationJson.has("slabHalf")) {
				val slab: Slab = placedBlockData as Slab
				slab.type = Type.valueOf(blockInformationJson.get("slabHalf").asString)
				placedBlockData = slab
			}

			if (blockInformationJson.has("persistent")) {
				val leaves: Leaves = placedBlockData as Leaves
				leaves.isPersistent = blockInformationJson.get("persistent").asBoolean
				placedBlockData = leaves
			}

			placedBlockState.blockData = placedBlockData
			placedBlockState.update()
		}
	}

}